v2/controller.go

```diff
@@ -0,0 +1,212 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package v2
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	statsv2 "github.com/containerd/cgroups/v2/stats"
+	"github.com/pkg/errors"
+)
+
+type cgValuer interface {
+	Values() []CGStr
+}
+
+type Resources struct {
+	CPU    *CPU
+	Memory *Memory
+	Pids   *Pids
+}
+
+func (r *Resources) Values() (o []CGStr) {
+	values := []cgValuer{
+		r.CPU,
+		r.Memory,
+		r.Pids,
+	}
+	for _, v := range values {
+		if v == nil {
+			continue
+		}
+		o = append(o, v.Values()...)
+	}
+	return o
+}
+
+type CGStr struct {
+	key   string
+	value interface{}
+}
+
+func (c *CGStr) write(path string, perm os.FileMode) error {
```

Comment:
please add a code comment to clarify `path` is file path (including `/sys/fs/cgroup` prefix), not a group path?